### DISCUSSION TITLE
Flatline rule wrongly triggered

### DISCUSSION THREAD

**Discussion by fberrez** [jertel/elastalert2#1191] [Q&A]
I have the same issue. Here is my rule:

```yml
name: myRule
description: Sends alerts on flatline
index: myIndex
type: flatline
use_count_query: true
timeframe:
  minutes: 5
threshold: 1

filter:
  - query:
      query_string:
        query: 'fields.metadata.entity: myEntity AND fields.metadata.code: myCode'

# Sets a time interval between two alerts of a same entity
realert:
  minutes: 1
```

And my kibana result: 
![image](https://github.com/jertel/elastalert2/assets/25006376/d5a73ec0-d1ed-425d-b1d4-ac008494ce6b)

We can see that I have one log every single minute. Meanwhile, the alert is triggered (flatline less than 1 in a timeframe of 5 minutes). The interesting thing is the alert is triggered when I set my elastalert config as:

```yml
run_every:
  minutes: 1

buffer_time:
  minutes: 10
```

But the alert will be triggered only if the flatline is confirmed (i.e no log at all in the last 5 minutes, the rule works as expected), when I change it to:

```yml
run_every:
  minutes: 3

buffer_time:
  minutes: 10
```

# expected behavior

1. With the config `run_every: minutes: 1`

- There is 1 hit on the first run (00:00)
- 1 hit on the second run (00:01)
- 1 hit on the third run (00:02)
- 1 hit on the fourth run (00:03)
- 1 hit on the fifth run (00:04)

5 hits on a `timeframe` of 5 minutes so no flatline (number of hits < 1) -> alert is not triggered

2. With the config `run_every: minutes: 3`

- There is 1 hit on the first run (00:00)
- 3 hits on the second run (00:03)
- 3 hits on the third run (00:06)

5 hits on a `timeframe` of 5 minutes so no flatline (number of hits < 1) -> alert is not triggered

**Comment by jertel** [comment]
I'm unable to reproduce this. 

The following rule and config does not trigger an alert, as expected:

```
name: flatline1
type: flatline
index: frequent
filter:
- query:
    query_string:
      query: 'status: alive'
threshold: 1
timeframe:
  minutes: 5
alert:
- debug
realert:
  minutes: 1
```

```
run_every:
  minutes: 1

buffer_time:
  minutes: 10
```


However, this rule and config does, correctly, trigger an alert:

```
name: flatline2
type: flatline
index: frequent
filter:
- query:
    query_string:
      query: 'status: alive'
threshold: 1
timeframe:
  seconds: 30
alert:
- debug
realert:
  minutes: 1
```

```
run_every:
  seconds: 10

buffer_time:
  minutes: 10
```

Here is the indexed data:
![image](https://github.com/jertel/elastalert2/assets/16639611/5f39f2a9-b4ca-4e85-80d8-e431a4e39eab)

I'm also attaching the CSV I used to create that index.
[frequent.csv](https://github.com/jertel/elastalert2/files/11671120/frequent.csv)

And the debug log showing the details of what's going on for the first 6+ minutes.
[frequent.log](https://github.com/jertel/elastalert2/files/11675937/frequent.log)

**Reply by fberrez** [reply]
Thank you for trying. But for me it really looks like an unexpected behavior because:

- it occurs at some random moments
- when it occurs, I restart my `elastalert` container and it works once again

I don't know how I can give you more context about this issue. Do you need more details? Which ones?

**Reply by jertel** [reply]
I suggest you compare the debug log I attached with your debug output. Observe differences that lead to your installation triggering alerts.

**Reply by fberrez** [reply]
I've never figured out how to run cli `elastalert` on a Mac M1

**Reply by jertel** [reply]
You can use the container to generate the debug logs.

**Reply by fberrez** [reply]
Just a quick message to say that I didn't forgot to try things with this rule. I want to understand why it is randomly triggered. And the logs say nothing interesting for now.

**Reply by fberrez** [reply]
This morning, the alert is triggered even if I have my logs in my elasticsearch (i.e no flatline at all)
I have these logs related to this alert:

```
INFO:elastalert:Queried rule (myRule) myRule - Logs flatline from 2023-06-16 08:36 UTC to 2023-06-16 08:37 UTC: 0 hits
INFO:elastalert:Queried rule (myRule) myRule - Logs flatline from 2023-06-16 08:37 UTC to 2023-06-16 08:37 UTC: 0 hits
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ran (myRule) myRule - Logs flatline from 2023-06-16 08:36 UTC to 2023-06-16 08:37 UTC: 0 query hits (0 already seen), 4 matches, 0 alerts sent
INFO:elastalert:(myRule) myRule - Logs flatline range 61

...

INFO:elastalert:Queried rule (myRule) myRule - Logs flatline from 2023-06-16 08:37 UTC to 2023-06-16 08:38 UTC: 0 hits
INFO:elastalert:Queried rule (myRule) myRule - Logs flatline from 2023-06-16 08:38 UTC to 2023-06-16 08:38 UTC: 0 hits
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ran (myRule) myRule - Logs flatline from 2023-06-16 08:37 UTC to 2023-06-16 08:38 UTC: 0 query hits (0 already seen), 4 matches, 0 alerts sent
INFO:elastalert:(myRule) myRule - Logs flatline range 62

...

INFO:elastalert:Queried rule (myRule) myRule - Logs flatline from 2023-06-16 08:38 UTC to 2023-06-16 08:39 UTC: 0 hits
INFO:elastalert:Queried rule (myRule) myRule - Logs flatline from 2023-06-16 08:39 UTC to 2023-06-16 08:39 UTC: 0 hits
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ignoring match for silenced rule (myRule) myRule - Logs flatline.all
INFO:elastalert:Ran (myRule) myRule - Logs flatline from 2023-06-16 08:38 UTC to 2023-06-16 08:39 UTC: 0 query hits (0 already seen), 4 matches, 0 alerts sent
INFO:elastalert:(myRule) myRule - Logs flatline range 63
```

**Reply by jertel** [reply]
The log shows "0 alerts sent" but it seems like you're saying it did send an alert. Perhaps you have another ElastAlert 2 instance running somewhere?

**Reply by fberrez** [reply]
Ok now I have this log: 
`INFO:elastalert:Ran (PROJECT) name - Logs flatline from 2023-06-27 11:30 UTC to 2023-06-27 11:31 UTC: 0 query hits (0 already seen), 4 matches, 1 alerts sent`

What does `4 matches` mean?

<img width="1280" alt="image" src="https://github.com/jertel/elastalert2/assets/25006376/4f28509b-01fc-4354-9c36-2cc7e2eee86d">

I tried different config but it doesn't help. Here is the last update:

```yml
name: (PROJECT) name - Logs flatline
description: Sends alerts when there is a flatline on the project logs
index: project-*
type: flatline
use_count_query: true
timeframe:
  minutes: 80
threshold: 1

filter:
  - query:
      query_string:
        query: '_id: *'

# Sets a time interval between two alerts of a same entity
realert:
  minutes: 5

start_time: 'SERVICE_ALERT_START_TIME'
end_time: 'SERVICE_ALERT_END_TIME'
drop_if: 'outside'

match_enhancements:
  - 'elastalert_modules.hour_range_enhancement.HourRangeEnhancement'

# Defines alert output
alert:
  - 'slack'

slack_webhook_url: https://.....
slack_username_override: elastalert

# Builds alert message
alert_text_type: alert_text_only

alert_subject: '[PROJECT] Flatline'

alert_text: 'Project did not send logs for the last 80 minutes'
```

**Reply by jertel** [reply]
matches = the total matches the rule has found since the start of the timeframe

**Reply by fberrez** [reply]
So `4 matches` have been found, the rule shouldn't be triggered since the flatline threshold is at `1`.

**Reply by jertel** [reply]
Matches are created when the threshold has been breached.

**Reply by fberrez** [reply]
Well at the end I found another solution than using flatline rules.
